File: solutions/java/collatz-conjecture/1/src/main/java/CollatzCalculator.java
class CollatzCalculator {

    int computeStepCount(int start) {
        if (start < 1) {
            throw new IllegalArgumentException("Only positive integers are allowed");
        }
        return computeStepCountR(start, 0);
    }

    private int computeStepCountR(int n, int stepCount) {
        if (n == 1) {
            return stepCount;
        } else if (n % 2 == 0) {
            return computeStepCountR(n / 2, stepCount + 1);
        } else {
            return computeStepCountR(n * 3 + 1, stepCount + 1);
        }
    }

}

File: solutions/kotlin/collatz-conjecture/1/src/main/kotlin/CollatzCalculator.kt
object CollatzCalculator {
    fun computeStepCount(start: Int): Int = computeStepCountR(start, 0)

    fun computeStepCountR(n: Int, stepCount: Int): Int = if (n < 1)  throw IllegalArgumentException("positive integers only") 
    else if (n == 1) stepCount 
    else if (n.mod(2) == 0) computeStepCountR(n.div(2), stepCount + 1) 
    else computeStepCountR(n.times(3).plus(1), stepCount + 1)
}
